File: apps/android/app/src/main/java/com/exponential/app/domain/MergeFailure.kt
package com.exponential.app.domain

import com.exponential.app.data.api.isConflictError
import com.exponential.app.data.api.trpcErrorMessage

/**
 * EXP-533: why a squash merge was refused, and whether the builtin "Fix merge
 * conflicts" recovery run could help.
 *
 * A merge fails for several unrelated reasons — a real content conflict,
 * branch protection, a stale base, an unconfigured GitHub App, or simply no
 * network — and only the FIRST is something a rebase-and-force-push run can
 * resolve. Offering the run for the others (the issue's complaint: it showed
 * up after a failed merge with no internet at all) sends the user off to burn
 * an agent session on a problem it cannot see.
 */
data class MergeFailure(
    /** The caption rendered on the failing row — already user-presentable. */
    val message: String,
    /** The server answered a real content conflict (HTTP 409). */
    val isConflict: Boolean,
) {
    companion object {
        fun from(error: Throwable, fallback: String): MergeFailure = MergeFailure(
            message = trpcErrorMessage(error, fallback),
            isConflict = isConflictError(error),
        )
    }
}

/**
 * Whether a failed merge may offer the "Fix merge conflicts" run: a real
 * conflict, on a PR whose branch we recorded (the run rebases it), with remote
 * start available.
 *
 * [steerEnabled] defaults to true for the surfaces that carry no separate
 * remote-start gate of their own (Reviews rows, which report an unreachable
 * device through the start sheet instead).
 */
fun canOfferFixConflicts(
    failure: MergeFailure?,
    branch: String?,
    steerEnabled: Boolean = true,
): Boolean = failure?.isConflict == true && !branch.isNullOrBlank() && steerEnabled

File: apps/web/src/lib/merge-failure.ts
import { trpcErrorCode, trpcErrorMessage } from "@/lib/trpc-error"

/**
 * EXP-533: what a refused merge means for the UI. `conflict` is the ONE gate
 * on the "Fix conflicts" recovery run: that run rebases, resolves and merges,
 * which only helps when the two trees actually disagree. A stale base, branch
 * protection, an unreachable server or a squash policy all refuse the merge
 * too, and offering the button there sends the agent in circles.
 */
export interface MergeFailure {
  message: string
  conflict: boolean
}

export function mergeFailure(error: unknown, fallback: string): MergeFailure {
  return { message: trpcErrorMessage(error, fallback), conflict: isConflict(error) }
}

function isConflict(error: unknown): boolean {
  // A REAL content conflict is the server's CONFLICT (409); every other
  // refusal (stale base, branch protection, policy) is PRECONDITION_FAILED.
  return trpcErrorCode(error) === `CONFLICT`
}
